### PR TITLE
inclusion of Function.Structures.BiInverse

### PR DESCRIPTION
**Pull request by pnlph**
New records BiEquivalence and BiInverse in Function.Bundles
New record IsBiInverse in Function.Structures
#1055

### CODE REVIEW

File: src/Function/Bundles.agda [hunk 1]
```diff
@@ -189,6 +189,51 @@ module _ (From : Setoid a ℓ₁) (To : Setoid b ℓ₂) where
...
 
 
+  record BiEquivalence : Set (a ⊔ b ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      f     : A → B
+      g₁    : B → A
+      g₂    : B → A
+      cong₁ : f Preserves _≈₁_ ⟶ _≈₂_
+      cong₂ : g₁ Preserves _≈₂_ ⟶ _≈₁_
+      cong₃ : g₂ Preserves _≈₂_ ⟶ _≈₁_
```
Comment (jespercockx) [inline src/Function/Bundles.agda]: Maybe there could be some more descriptive names for these fields, e.g. `cong-f`, `cong-g₁`, and `cong-g₂`?

File: src/Function/Bundles.agda [hunk 2]
```diff
@@ -189,6 +189,51 @@ module _ (From : Setoid a ℓ₁) (To : Setoid b ℓ₂) where
...
+
+
+  record BiInverse : Set (a ⊔ b ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      f         : A → B
+      g₁        : B → A
+      g₂        : B → A
+      cong₁     : f Preserves _≈₁_ ⟶ _≈₂_
+      cong₂     : g₁ Preserves _≈₂_ ⟶ _≈₁_
+      cong₃     : g₂ Preserves _≈₂_ ⟶ _≈₁_
```
Comment (jespercockx) [inline src/Function/Bundles.agda]: Same here.

File: src/Function/Bundles.agda (hunk 1 above)
Comment (pnlph) [reply src/Function/Bundles.agda]: I agree, that notation is more descriptive and clear, maybe @MatthewDaggitt can say something about the impact of this change. In order to preserve consistency, `cong₁` and `cong₂` should be also elsewhere in the library replaced by `cong-f` and `cong-g` respectively.

Comment (gallais) [reply src/Function/Bundles.agda]: I believe we usually name things `f-cong` and use `Congruent`.
Cf. [`Algebra.Structures`](https://agda.github.io/agda-stdlib/Algebra.Structures.html)

File: src/Function/Bundles.agda [hunk 3]
```diff
@@ -189,6 +189,51 @@ module _ (From : Setoid a ℓ₁) (To : Setoid b ℓ₂) where
...
+      inverseˡ  : Inverseˡ f g₁
+      inverseʳ  : Inverseʳ f g₂
+
+    isCongruent : IsCongruent f
```
Comment (MatthewDaggitt) [inline src/Function/Bundles.agda]: I think this needs to be `f-isCongruent` as it is also possible to show `g₁-isCongruent` and `g₂-isCongruent`

File: src/Function/Structures.agda [hunk 4]
```diff
@@ -100,6 +100,19 @@ record IsRightInverse (f : A → B) (g : B → A) : Set (a ⊔ b ⊔ ℓ₁ ⊔
     renaming (cong to cong₁)
 
 
+record IsBiInverse
+  (f : A → B) (g₁ : B → A) (g₂ : B → A) : Set (a ⊔ b ⊔ ℓ₁ ⊔ ℓ₂) where
+  field
+    isCongruent : IsCongruent f
```
Comment (MatthewDaggitt) [inline src/Function/Structures.agda]: Same here

**Comment by pnlph** [comment]
Thanks, minor changes sent.

Just an observation: when editing the changelog, i think i introduced a newline at the end of the file. I see that Atom does this by default, so I found [why should text files end with a newline](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline).

The reason seems to be that this is how [the POSIX standard defines a line](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206).

File: CHANGELOG.md [hunk 5]
```diff
@@ -675,4 +682,4 @@ Version 2.6.1 changes
 * Added new definitions in `Relation.Binary.Core`:
   ```agda
   DecidableEquality A = Decidable {A = A} _≡_
-  ```
\ No newline at end of file
+  ```  ```
```
Comment (gallais) [inline CHANGELOG.md]: duplicated backtick guards?

File: src/Function/Bundles.agda [hunk 6]
```diff
@@ -189,6 +189,51 @@ module _ (From : Setoid a ℓ₁) (To : Setoid b ℓ₂) where
       }
 
 
+  record BiEquivalence : Set (a ⊔ b ⊔ ℓ₁ ⊔ ℓ₂) where
```
Comment (MatthewDaggitt) [inline src/Function/Bundles.agda]: Hmm is this definition needed or not? If it is then we should make sure we add a `IsBiEquivalence` record as well and make sure it's added to the CHANGELOG.

Comment (pnlph) [reply src/Function/Bundles.agda]: OK, but `IsEquivalence` is is not a record itself, but defined as a field in the `Setoid` record.

Should we better add a `BiSetoid` record to `Relation/Binary/Bundles.agda` that includes a `IsBiEquivalence` field and the arrow symbol `_↩↪_` that you proposed in [574466050](https://github.com/agda/agda-stdlib/issues/1055#issuecomment-574466050)?

Comment (MatthewDaggitt) [reply src/Function/Bundles.agda]: > OK, but IsEquivalence is not a record itself, but defined as a field in the Setoid record.

I'm not quite sure what you mean by this. `IsEquivalence` is a record:
https://github.com/agda/agda-stdlib/blob/15dc42e1eeb0dbf946961213ffd86b5fa60ccb00/src/Relation/Binary/Structures.agda#L47

> Should we better add a BiSetoid record to Relation/Binary/Bundles.agda that includes a IsBiEquivalence field and the arrow symbol _↩↪_ that you proposed in 574466050?

No I don't think so. A `BiEquivalence` is still talking about functions between two types and therefore belongs in the function hierarchy. The `_↩↪_` symbol I was referring to, should be used for the instantiation of `BiEquivalence` with propositional equality. See the examples at the bottom of `Function.Bundles`.

Comment (pnlph) [reply src/Function/Bundles.agda]: > I'm not quite sure what you mean by this. `IsEquivalence` is a record:

Sorry, when retrieving the `agda-mode:go-to-definition` command, I mixed `isEquivalence` with `IsEquivalence`. Of course, you are right, `IsEquivalence` is a record in `Relation/Binary/Structures.agda`

Thank you for the hint, I added a new BiInverse bundle `_↩↪_` and a new constructor `mk↩↪`  following the notation of the examples at the end of `Funtion.Bundles`

**Pull request merged**